replica/database.cc

```diff
@@ -1647,6 +1648,33 @@ database::query_mutations(schema_ptr query_schema, const query::read_command& cm
     co_return std::tuple(std::move(result), hit_rate);
 }
 
+db::view::update_backlog database::get_view_update_backlog() const {
+    size_t current = 0;
+    size_t max = 1;
+    auto update_backlog = [&current, &max] (size_t new_current, size_t new_max) {
+        if (new_max == 0) {
+            return;
+        }
+        if ((double)new_current / new_max > (double)current / max) {
+            current = new_current;
+            max = new_max;
+        }
+    };
+    const reader_concurrency_semaphore* sem_ptr = _view_update_read_concurrency_semaphores_group.get_or_null(seastar::current_scheduling_group());
+    // First calculate the pressure on reader_concurrency_semaphore
+    if (sem_ptr) {
+        // Check the read queue size
+        update_backlog(sem_ptr->get_stats().waiters, sem_ptr->get_max_queue_length());
+        // Check the memory usage
+        update_backlog(sem_ptr->consumed_resources().memory, sem_ptr->get_kill_limit());
```

Comment:
Queued reads don't consume memory. The memory consumption solely depends on the activities of the currently admitted reads. This is influenced by things like data shape, sstable layout as well as memtable/cache content.

All that said, it may make sense to include this in the backlog calculation. Memory usage of reads may indicate amount of work that needs to be done by future reads (velocity).

Reply:
> Queued reads don't consume memory. The memory consumption solely depends on the activities of the currently admitted reads. This is influenced by things like data shape, sstable layout as well as memtable/cache content.

Right, the `get_max_queue_length()` part is for queued reads and the `get_kill_limit()` is for admitted reads (though this one is very unlikely to get hit, due to a limited number of admitted reads)
Although from 
```c++
// Assume a queued read takes up 1kB of memory, and allow 2% of memory to be filled up with such reads.
size_t max_inactive_queue_length() { return _dbcfg.available_memory * 0.02 / 1000; }
```
 looks like the queued reads also take a bit of memory, even though it's not tracked by the semaphore

Reply:
This comment is out of date, it was written in a time, where readers were queued on the semaphore *after* they were created. Nowadays we first get the permit through admission and *only then* create the reader.
The permit itself of course consumes some memory, but nothing close to 1kb, more like a few dozen bytes.